datalad_next/url_operations/ssh.py

```diff
@@ -81,63 +93,56 @@ def stat(self,
-
-        ssh_cat = _SshCat(url)
-        stream = ssh_cat.run(cmd, protocol=StdOutCaptureGeneratorProtocol)
-        for chunk in stream:
-            if need_magic:
-                expected_magic = need_magic[:min(len(need_magic),
-                                                 len(chunk))]
-                incoming_magic = chunk[:len(need_magic)]
-                # does the incoming data have the remaining magic bytes?
-                if incoming_magic != expected_magic:
-                    raise RuntimeError(
-                        "Protocol error: report header not received")
-                # reduce (still missing) magic, if any
-                need_magic = need_magic[len(expected_magic):]
-                # strip magic from input
-                chunk = chunk[len(expected_magic):]
-            if chunk and expected_size is None:
-                # we have incoming data left and
-                # we have not yet consumed the size info
-                size_data = chunk.split(b'\1', maxsplit=1)
-                expected_size_str += size_data[0]
-                if len(size_data) > 1:
-                    # this is not only size info, but we found the start of
-                    # the data
-                    expected_size = int(expected_size_str)
-                    chunk = size_data[1]
-            if expected_size:
-                props = {
-                    'content-length': expected_size,
-                    '_stream': chain([chunk], stream) if chunk else stream,
-                }
-                return props
-            # there should be no data left to process, or something went wrong
-            assert not chunk
+        # b'\1', and the file content after that.
+        magic_marker = b'\1\2\3'
+
+        # use the `align_pattern` iterable to guarantees, that the magic
+        # marker is always contained in a complete chunk.
+        # TODO: the align-pattern iterator stays "on" the stream, which might
+        #  slow down processing upstream. We could consider to change
+        #  align_pattern to an iterator-class and support detaching.
+        aligned_stream = align_pattern(stream, magic_marker)
+
+        # Because the stream should start with the pattern, the first chunk should contain it.
+        chunk = next(aligned_stream)
+        if chunk[:len(magic_marker)] != magic_marker:
+            raise RuntimeError("Protocol error: report header not received")
+        chunk = chunk[len(magic_marker):]
+
+        # The length is transferred now and terminated by b'\x01'.
+        while b'\x01' not in chunk:
+            chunk += next(stream)
+
+        marker_index = chunk.index(b'\x01')
+        expected_size = int(chunk[:marker_index])
+        chunk = chunk[marker_index + 1:]
+        props = {
+            'content-length': expected_size,
+            '_stream':
+                chain([chunk], aligned_stream)
+                if chunk
+                else aligned_stream
```

Comment:
It there anything that would speak against the following:

```suggestion
                chain([chunk], aligned_stream)
                if chunk
                else aligned_stream
```

The `align_patter` iterator advanced the `stream` enough to get to the file content. The will be no other marker to look for. We should be able to detach and go back to the original iterable.

There is a comment a few lines above:

> # TODO: the align-pattern iterator stays "on" the stream, which might
> #  slow down processing upstream. We could consider to change
> #  align_pattern to an iterator-class and support detaching.

This is talking about the same issue, but suggests a different,
seemingly more sophisticated approach.

@christian-monch can you comment on the adequacy of my proposal?

Reply:
This is done now.

Reply:
Tests remain happy.